### PR TITLE
January release notes for developer framework

### PR DESCRIPTION
**Pull request by nrichers**
## Internal Notes for Reviewers

This PR includes: 

- January release notes for the developer framework
- Updated copyright notice for 2024
- Updated Makefile to generate test descriptions on the fly (looks like I'd missed this in https://github.com/validmind/documentation/pull/144), plus quietened output
- Refreshed notebooks and developer framework HTML docs 
- Updated the supported Python versions (and switched to a variable to make future updates easier)

Open items:

- [ ] Looks like we might have a couple of new tests without release notes? 

   image

- [x] https://github.com/validmind/developer-framework/pull/316 did a really nice job cleaning up our notebooks, but we need to figure out how to sweat more of the details. E.g. this new notebook that gets linked to from the release notes has some clearly unfinished stuff right at the top that both the original author and reviewer missed: 

   image

## Output

Capto_Capture 2024-01-24_01-52-20_pm

### CODE REVIEW

File: site/releases/2024-jan-22/highlights.qmd [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+results = vm.run_documentation_tests(
+    section=["data_preparation", "model_development"],
+    dataset=vm_dataset,
+)
+```
+
+Try it yourself: [Running individual documentation sections](../../notebooks/how_to/run_documentation_sections.ipynb)
+
+<!---feat: [sc-2468] Allow arbitrary test context (V2) by @johnwalz97 in [#312](https://github.com/validmind/developer-framework/pull/312)--->
+#### Support for arbitrary and custom inputs <span id="arbitrary-test-new">
```
Comment (nrichers) [inline site/releases/2024-jan-22/highlights.qmd:103]: Can we just say "custom inputs" to simplify the text? That seems like it might entail arbitrary ones unless that is a specific keyword we need to cover.

Comment (mehdi0501) [reply site/releases/2024-jan-22/highlights.qmd:107]: agree

File: site/releases/2024-jan-22/highlights.qmd [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+     feature_columns=[...],
+     target_column= ...,
+     extra_columns={
+         prediction_column: 'NAME-OF-PREDICTION-COLUMN',
+    },
+)
+```
+
+<!---[SC-2438] Support for adding feature columns parameter to init dataset by @juanmleng in [#279](https://github.com/validmind/developer-framework/pull/279)--->
+#### Improvements to `init dataset`
```
Comment (mehdi0501) [inline site/releases/2024-jan-22/highlights.qmd:51]: I believe it should be "init_dataset" (not "init dataset")

File: site/releases/2024-jan-22/highlights.qmd [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+
+This release includes numerous improvements to the ValidMind Developer Framework, including new features for model and dataset initialization, easier testing, support for additional inputs and the Azure OpenAI API, updated notebooks, bug fixes, deprecations, and much more.
+
+### ValidMind Developer Framework (v1.25.3)
+
+<!---[SC-2628] `init_model` improvements  by @AnilSorathiya in [#304](https://github.com/validmind/developer-framework/pull/304)--->
+
+#### Improvements to `init_model`
+
+When initializing a model, you can now pass a dataset with pre-computed model predictions to improve performance. By default, if no prediction column is specified when calling `init_model`, the ValidMind Developer Framework will compute the model predictions as before.
```
Comment (mehdi0501) [inline site/releases/2024-jan-22/highlights.qmd:16]: this description is a bit unclear to me. 
What does "compute model predictions as before" refer to?
What is meant by "improve performance"?

File: site/releases/2024-jan-22/highlights.qmd [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+feature_columns = ['CreditScore', 'Age', 'Balance', 'NumOfProducts', 'EstimatedSalary']
+
+vm_train_ds = vm.init_dataset(
+    dataset=train_df,
+    target_column=demo_dataset.target_column,
+    feature_columns=feature_columns
+)
+```
+
+Additionally, two new metrics enable you to report by segment and show the use of `feature_columns`:
```
Comment (mehdi0501) [inline site/releases/2024-jan-22/highlights.qmd:69]: I think the two new metrics deserve their own highlight - @cachafla  agree?

File: site/releases/2024-jan-22/highlights.qmd [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+A new notebook illustrates how you can configure these dataset features:
+
+- How to utilize the `feature_columns` parameter when initizalizing `validmind` datasets and model objects
+- How `feature_columns` can be used to report by segment
+
+Try it: [Configuring and Using Dataset Features](../../notebooks/how_to/configure_dataset_features.ipynb)
+
+<!---[SC-2707] Support for multiple sections in run_documentation_tests() by @juanmleng in [#307](https://github.com/validmind/developer-framework/pull/307)--->
+#### Improvements to `run_documentation_tests()`
```
Comment (mehdi0501) [inline site/releases/2024-jan-22/highlights.qmd:82]: description in this highlight is unclear to me. 
- what is the benefit for the user of running multiple sections at a time?
- I don't think we need to specify that this came through a customer feedback (makes it sound like the other improvements are not customer driven)

File: site/releases/2024-jan-22/highlights.qmd (hunk 1 above)
Comment (mehdi0501) [reply site/releases/2024-jan-22/highlights.qmd:107]: @cachafla  what is the difference between "arbitrary" and "custom" inputs?

File: site/releases/2024-jan-22/highlights.qmd (hunk 2 above)
Comment (nrichers) [reply site/releases/2024-jan-22/highlights.qmd:51]: Good catch! Fixed.

File: site/releases/2024-jan-22/highlights.qmd (hunk 3 above)
Comment (nrichers) [reply site/releases/2024-jan-22/highlights.qmd:16]: I think these would mainly be questions for @AnilSorathiya since it's from his [#304](https://github.com/validmind/developer-framework/pull/304). Anil, would you be able to provide some answers to Mehdi's questions, please?

File: site/releases/2024-jan-22/highlights.qmd (hunk 4 above)
Comment (nrichers) [reply site/releases/2024-jan-22/highlights.qmd:69]: Is there a place where we could show more information about metrics? If I search for `MissingValuesBySegment` in the docs site, the only result I get is in these release notes. Meaning, if we make these metrics their own highlight, there is nowhere to link to for more information.

File: site/releases/2024-jan-22/highlights.qmd (hunk 5 above)
Comment (nrichers) [reply site/releases/2024-jan-22/highlights.qmd:82]: > * what is the benefit for the user of running multiple sections at a time?

When I was working on these release notes, I traced this back to https://app.shortcut.com/validmind/story/2707/run-documentation-tests-should-run-for-multiple-sections, but that's all the info there is. It seem mainly a convenience feature ("...  it would be beneficial ...". 

> * I don't think we need to specify that this came through a customer feedback (makes it sound like the other improvements are not customer driven)

Removed.

File: site/releases/2024-jan-22/highlights.qmd (hunk 3 above)
Comment (cachafla) [reply site/releases/2024-jan-22/highlights.qmd:16]: > What is meant by "improve performance"?

We can reword this to communicate that users can save time when they have big datasets and (probably) have pre-computed model predictions somewhere.

```
When initializing a model, you can now pass a dataset with pre-computed model predictions if they are available.
```

> What does "compute model predictions as before" refer to?

Every time we call `init_model()` we run predictions on the entire dataset.

Suggestion: just remove the word "before":

```
By default, if no prediction column is specified when calling `init_model`, the ValidMind Developer Framework will compute the model predictions on the entire dataset.
```

File: site/releases/2024-jan-22/highlights.qmd [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+feature_columns = ['CreditScore', 'Age', 'Balance', 'NumOfProducts', 'EstimatedSalary']
+
+vm_train_ds = vm.init_dataset(
+    dataset=train_df,
+    target_column=demo_dataset.target_column,
+    feature_columns=feature_columns
+)
+```
+
+Additionally, two new metrics enable you to report by segment and show the use of `feature_columns`:
```
Comment (cachafla) [inline site/releases/2024-jan-22/highlights.qmd:69]: Any metric that uses a dataset is affected by custom feature columns :).

```suggestion
```

File: site/releases/2024-jan-22/highlights.qmd (hunk 5 above)
Comment (cachafla) [reply site/releases/2024-jan-22/highlights.qmd:82]: > what is the benefit for the user of running multiple sections at a time?

The user doesn't need to call the same function twice for two different sections, resulting in code that is less prone to errors: less duplication and ability to use a single config object. Something like this:

## Before

```python
full_suite = vm.run_documentation_tests(
    inputs = {
        ...
    },
    section="section_1",
    config={
        "validmind.tests.data_validation.ClassImbalance": ...
    } 
)

full_suite = vm.run_documentation_tests(
    inputs = {
        ...
    },
    section="section_2",
    config={
        "validmind.tests.data_validation.Duplicates": ...
    } 
)
```

## After

Single function call:

```python
full_suite = vm.run_documentation_tests(
    inputs = {
        ...
    },
    section=["section_1", "section_2"],
    config={
        "validmind.tests.data_validation.ClassImbalance": ...,
        "validmind.tests.data_validation.Duplicates": ...
    } 
)
```

File: site/releases/2024-jan-22/highlights.qmd (hunk 1 above)
Comment (cachafla) [reply site/releases/2024-jan-22/highlights.qmd:107]: The same. We can just use custom inputs.

File: site/releases/2024-jan-22/highlights.qmd (hunk 3 above)
Comment (AnilSorathiya) [reply site/releases/2024-jan-22/highlights.qmd:16]: > this description is a bit unclear to me.
> What does "compute model predictions as before" refer to?
> What is meant by "improve performance"?

Currently, we run prediction on given datasets (training, test datasets) while calling `init_model` method and store prediction values in the model object (vm_model).  Therefore it takes time to create vm_model object. 
Now this functionality provides flexibility to pass `prediction values`  as a additional column in the datasets while creating vm_model object. If prediction column is provided, we don't need to run model prediction while creating `vm_mdoel`.  This means it's going to speed up.

There is another reason to do this as well. Sometime model developer wants to compute prediction while creating a model for just once. The reason could be datasets are very large. In this case he might perform prediction out side of VM env. 

If prediction column is not provided, it will follow current behaviour.

File: site/releases/2024-jan-22/highlights.qmd (hunk 1 above)
Comment (nrichers) [reply site/releases/2024-jan-22/highlights.qmd:107]: Updated globally to refer only to "custom inputs".

File: site/releases/2024-jan-22/highlights.qmd (hunk 3 above)
Comment (nrichers) [reply site/releases/2024-jan-22/highlights.qmd:16]: Thank you both, @AnilSorathiya and @cachafla! I have updated the text as suggested by Andres. (If you would like me to include more of Anil's added explanation let me know. I found it helpful to understand the feature better but the gist seems to be conveived by Andres' rewording.)

File: site/releases/2024-jan-22/highlights.qmd (hunk 5 above)
Comment (nrichers) [reply site/releases/2024-jan-22/highlights.qmd:82]: Thank you, this is very helpful. I wove that explanation into the text and included the before/after example usage to illustrate.

File: site/releases/2024-jan-22/highlights.qmd (hunk 6 above)
Comment (nrichers) [reply site/releases/2024-jan-22/highlights.qmd:69]: I removed this manually as we also had a couple of actual metrics to remove right below.

File: site/releases/2024-jan-22/highlights.qmd (hunk 4 above)
Comment (nrichers) [reply site/releases/2024-jan-22/highlights.qmd]: Andres suggested to remove this text.

**Comment by nrichers** [comment]
@mehdi0501 and @cachafla I have resolved your comments and updated the root comment with the latest output. Could you take another look and let me know if the release notes are ready? (We probably need to update the release date again but I will wait until the release notes are otherwise ready.)

**Review by mehdi0501 — APPROVED** [review]
Changes look good to me, thank you for addressing the comments around explanations of the new dev framework features!

**Review by cachafla — APPROVED** [review]
Great 👏

**Comment by nrichers** [comment]
OK, updated the notebooks one last time and sanity checked that the rendered static site works as expected. Done, merging.
